pkg/verifier/safe_delete.go

```diff
@@ -14,10 +15,13 @@ import (
 	"github.com/pkg/errors"
 )
 
-// SafeDelete moves block to backup bucket and if succeeded, removes it from source bucket.
-// It returns error if block dir already exists in backup bucket (blocks should be immutable) or any
-// of the operation fails.
-func SafeDelete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, backupBkt objstore.Bucket, id ulid.ULID) error {
+// SafeDelete moves block to backup bucket and if succeeded, removes it from
+// source bucket.  It returns error if block dir already exists in backup
+// bucket (blocks should be immutable) or any of the operation fails.  If
+// the block has already been downloaded it must exist in tempdir, else it
+// will be downloaded to be copied to the backup bucket.  tempdir set to the
```

Comment:
`SafeDelete()` doesn't know the temporary directory where a previous function may have downloaded the block, so we need to pass in a path somehow.  Perhaps rename this argument something like `useExisting`?  Which would mean we use the existing block if found under that path, otherwise a new temporary directory is created and the block is downloaded.

Reply:
sorry I meant let's pass two parameters to explicitly describe the intention

Reply:
WIP

Reply:
Ok, committed a patch for makes this explicit.  I've tested the repair process with my collection of broken blocks I've started keeping and had the behavior expected.  (No repeated downloads.)

Reply:
Had thanos-compact cut a 450GiB TSDB block today which I referenced in https://github.com/improbable-eng/thanos/issues/1152 which is part of the motivation for this.